server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/ShardAssignment.java

```diff
@@ -30,7 +29,10 @@ public int assigned() {
 
     public static ShardAssignment ofAssignedShards(List<ShardRouting> routings) {
         assert routings.stream().allMatch(ShardRouting::started) : routings;
```

Comment:
Shard started assertion could be moved to the loop as well to avoid the second stream and speed up the test execution

Reply:
I don't think this shows up much in tests. Tests generally don't run over a large number of shards, so this kind of thing doesn't really have an impact on them. I benchmarked the test code here and there and the runtime of the allocation code was trivial :)